helix-term/src/commands/typed.rs

```
@@ -13,38 +13,82 @@ use helix_stdx::path::home_dir;
     fn completer_for_argument_number(&self, n: usize) -> &Completer {
-        match self.signature.positional_args.get(n) {
-            Some(completer) => completer,
-            _ => &self.signature.var_args,
+        self.signature
+            .completer
+            .positional_args
+            .get(n)
+            .unwrap_or(&self.signature.completer.var_args)
+    }
+
+    pub fn ensure_signature(&self, count: usize) -> anyhow::Result<()> {
+        match self.signature.positionals {
+            (0, Some(0)) => ensure!(count == 0, "`:{}` doesn't take any arguments", self.name),
+            (min, Some(max)) if min == max => ensure!(
+                (min..=max).contains(&count),
+                "`:{}` needs `{min}` argument{}, got {count}",
+                self.name,
+                if min > 1 { "'s" } else { "" }
+            ),
+            (min, Some(max)) if min == max => ensure!(
+                (min..=max).contains(&count),
+                "`:{}` needs at least `{min}` argument{} and at most `{max}`, got `{count}`",
+                self.name,
+                if min > 1 { "'s" } else { "" }
+            ),
+            (min, _) => ensure!(
+                (min..).contains(&count),
+                "`:{}` needs at least `{min}` argument{}",
+                self.name,
+                if min > 1 { "'s" } else { "" }
+            ),
         }
+
+        Ok(())
     }
 }
 
 #[derive(Clone)]
 pub struct CommandSignature {
+    // TODO: flags: flags![],
+    /// The min-max of the about of arguments a command can take.
+    ///
+    /// - **0**: (0, Some(0))
+    /// - **0-1**: (0, Some(1))
+    /// - **1**: (1, Some(1))
+    /// - **1-10**: (1, Some(10))
+    /// - **Unbounded**: (1, None)
+    positionals: (usize, Option<usize>),
+    pub parse_mode: ParseMode,
+    // TODO: better description with that explains more about the kinds of completers
+    /// What completion methods, if any, does this command have?
+    completer: CommandCompleter,
```

After thinking about this more I believe we should actually move this off of `CommandSignature` and back to a separate field on `TypableCommand`. The idea is that we should be able to pass the entire Signature type to Args so that it understands the number of expected parameters, parse mode and flags in the future.

The checking of min and max positionals can also be moved into `Args::from_signature` but it should also take a bool about whether to error which is set to true on `PromptEvent::Validate`. (See the constructor for `ParametersParser` in Kakoune.) We will want to do flag validation in that function later and will also need the ability to ignore errors like unknown flags.

---

The crate hierarchy means that it cant be passed in as a whole, should I move the signature to shellwords?

---

Yeah let's move the args stuff into a new `args` module in helix-core and define the `Signature` type (and later `Flags`) there too